xblock/scorable.py

```diff
@@ -0,0 +1,122 @@
+"""
+Scorable.
+"""
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+from collections import namedtuple
+import logging
+
+
+log = logging.getLogger(__name__)
+
+
+Score = namedtuple('Score', ['raw_earned', 'raw_possible'])
+
+
+class ScorableXBlockMixin(object):
+    """
+    Mixin to handle functionality related to scoring.
+
+    Subclasses must define the following:
+
+        ScorableXBlockMixin.has_score(self)
+        ScorableXBlockMixin.get_score(self)
+        ScorableXBlockMixin.set_score(self, score)
+        ScorableXBlockMixin.calculate_score(self, score)
+    """
+
+    def rescore(self, only_if_higher):
+        """
+        Calculate a new raw score and save it to the block.  If only_if_higher
+        is True and the score didn't improve, keep the existing score.
+
+        Raises a TypeError if the block cannot be scored.
+        Raises a ValueError if the user has not yet completed the problem.
+        """
+
+        _ = self.runtime.service(self, 'i18n').ugettext
+
+        if not self.allows_rescore():
+            raise TypeError(_('Problem does not support rescoring: {}').format(self.location))
+
+        if not self.has_score():
+            raise ValueError(_('Cannot rescore unanswered problem: {}').format(self.location))
+
+        try:
+            new_score = self.calculate_score()
+        except Exception as exc:  # pylint: disable=broad-except
+            return {
+                'success': False,
+                'error': exc.message,
+            }
+
+        self._publish_grade(new_score, only_if_higher)
+        return {
+            'success': True,
+            'correct': new_score.raw_earned == new_score.raw_possible,
+            'raw_earned': new_score.raw_earned,
+            'raw_possible': new_score.raw_possible,
+        }
+
+    def allows_rescore(self):
+        """
+        Boolean value: Can this problem be rescored?
+
+        Subtypes may wish to override this if they need conditional support for
+        rescoring.
+        """
+        return True
+
+    def has_score(self):
```

Comment:
This might collide with a current XBlock attribute that simply says whether the block _can_ have a score, not whether it does. It isn't well-documented in this repo but it's referred to here. https://github.com/edx/XBlock/blob/5b6af7a8cde09bd103ccd91919a858b6f088e28a/xblock/runtime.py#L507 

Drag and drop already sets a `has_score` attribute, and it's always `True`.